test_pipe_breaking_out_of_async_iter: Ignore <async_generator_athrow without __name__> task

The test scaffold that checks for orphan tasks may want to ignore the async_generator_athrow tasks, or just yield once to give them time to exit -- then check for orphan tasks again.


```
tests/test_shellous.py::test_pipe_breaking_out_of_async_iter 
----------------------------------------------------------------------------------------------------------------------------------- live log setup ------------------------------------------------------------------------------------------------------------------------------------
1636743144.724 DEBUG asyncio Using selector: EpollSelector
1636743144.725 DEBUG asyncio Using selector: EpollSelector
1636743144.726 DEBUG asyncio Close <_UnixSelectorEventLoop running=False closed=False debug=True>
------------------------------------------------------------------------------------------------------------------------------------ live log call ------------------------------------------------------------------------------------------------------------------------------------
1636743144.729 INFO shellous PipeRunner.__aenter__ stepin <PipeRunner 'env|tr'>
1636743144.730 INFO shellous PipeRunner._start stepin <PipeRunner 'env|tr'>
1636743144.731 INFO shellous PipeRunner._setup_capturing stepin <PipeRunner 'env|tr'>
1636743144.732 INFO shellous Runner.__aenter__ stepin <Runner 'env' pid=None> (Linux-5.4.0-90-generic-x86_64-with-glibc2.31 CPython 3.9.5 asyncio.unix_events._UnixSelectorEventLoop MainThread DefaultChildWatcher)
1636743144.733 INFO shellous Runner._start stepin <Runner 'env' pid=None>
1636743144.734 INFO shellous Runner._subprocess_spawn stepin <Runner 'env' pid=None>
1636743144.735 INFO shellous Runner.__aenter__ stepin <Runner 'tr' pid=None> (Linux-5.4.0-90-generic-x86_64-with-glibc2.31 CPython 3.9.5 asyncio.unix_events._UnixSelectorEventLoop MainThread DefaultChildWatcher)
1636743144.736 INFO shellous Runner._start stepin <Runner 'tr' pid=None>
1636743144.736 INFO shellous Runner._subprocess_spawn stepin <Runner 'tr' pid=None>
1636743144.739 INFO shellous Runner._subprocess_exec stepin <Runner 'env' pid=None>
1636743144.740 DEBUG asyncio execute program '/home/bfish/code/shellous/.venv/bin/python3.9' stdin=<pipe> stdout=9 stderr=<_io.TextIOWrapper name='<stderr>' mode='w' encoding='utf-8'>
1636743144.741 DEBUG shellous KQWorker starting <KQueueWorker(Thread-15, started daemon 140513983432448)>
1636743144.743 DEBUG shellous _add_read_event 11
1636743144.745 DEBUG asyncio process '/home/bfish/code/shellous/.venv/bin/python3.9' created: pid 36768
1636743144.746 DEBUG shellous watch_pid 36768
1636743144.747 INFO shellous Runner._subprocess_exec stepin <Runner 'tr' pid=None>
1636743144.748 DEBUG asyncio execute program '/home/bfish/code/shellous/.venv/bin/python3.9' stdin=8 stdout=<pipe> stderr=<_io.TextIOWrapper name='<stderr>' mode='w' encoding='utf-8'>
1636743144.748 DEBUG shellous _add_read_event 12
1636743144.755 DEBUG asyncio process '/home/bfish/code/shellous/.venv/bin/python3.9' created: pid 36769
1636743144.756 DEBUG shellous watch_pid 36769
1636743144.757 DEBUG shellous _add_read_event 16
1636743144.758 DEBUG asyncio Write pipe 13 connected: (<_UnixWritePipeTransport fd=13 idle bufsize=0>, <WriteSubprocessPipeProto fd=0 pipe=<_UnixWritePipeTransport fd=13 idle bufsize=0>>)
1636743144.759 DEBUG asyncio Read pipe 14 connected: (<_UnixReadPipeTransport fd=14 polling>, <ReadSubprocessPipeProto fd=1 pipe=<_UnixReadPipeTransport fd=14 polling>>)
1636743144.761 INFO asyncio execute program '/home/bfish/code/shellous/.venv/bin/python3.9': <_UnixSubprocessTransport pid=36768 running stdin=<_UnixWritePipeTransport fd=13 idle bufsize=0>>
1636743144.763 INFO shellous Runner._subprocess_exec stepout <Runner 'env' pid=36768 exit_code=None> ex=None
1636743144.767 INFO asyncio execute program '/home/bfish/code/shellous/.venv/bin/python3.9': <_UnixSubprocessTransport pid=36769 running stdout=<_UnixReadPipeTransport fd=14 polling>>
1636743144.771 INFO shellous Runner._subprocess_exec stepout <Runner 'tr' pid=36769 exit_code=None> ex=None
1636743144.772 INFO shellous Runner._subprocess_spawn stepout <Runner 'env' pid=36768 exit_code=None> ex=None
1636743144.773 INFO shellous Runner._start stepout <Runner 'env' pid=36768 exit_code=None> ex=None
1636743144.774 INFO shellous Runner.__aenter__ stepout <Runner 'env' pid=36768 exit_code=None> ex=None
1636743144.774 INFO shellous Runner.__aexit__ stepin <Runner 'env' pid=36768 exit_code=None> exc_value=None
1636743144.775 INFO shellous Runner._finish stepin <Runner 'env' pid=36768 exit_code=None>
1636743144.777 INFO shellous Runner._wait stepin <Runner 'env' pid=36768 exit_code=None>
1636743144.778 INFO shellous Runner._subprocess_spawn stepout <Runner 'tr' pid=36769 exit_code=None> ex=None
1636743144.783 INFO shellous Runner._start stepout <Runner 'tr' pid=36769 exit_code=None> ex=None
1636743144.787 INFO shellous Runner.__aenter__ stepout <Runner 'tr' pid=36769 exit_code=None> ex=None
1636743144.787 INFO shellous Runner.__aexit__ stepin <Runner 'tr' pid=36769 exit_code=None> exc_value=None
1636743144.788 INFO shellous Runner._finish stepin <Runner 'tr' pid=36769 exit_code=None>
1636743144.789 INFO shellous Runner._wait stepin <Runner 'tr' pid=36769 exit_code=None>
1636743144.790 INFO shellous write_stream stepin
1636743144.791 INFO shellous write_stream stepout ex=None
1636743144.791 INFO shellous Runner._waiter stepin <Runner 'env' pid=36768 exit_code=None>
1636743144.792 INFO shellous Runner._waiter stepin <Runner 'tr' pid=36769 exit_code=None>
1636743144.795 INFO shellous PipeRunner._setup_capturing stepout <PipeRunner 'env|tr'> ex=None
1636743144.798 INFO shellous os.waitpid returned (36768, 0)
1636743144.799 INFO shellous PipeRunner._start stepout <PipeRunner 'env|tr'> ex=None
1636743144.799 DEBUG shellous _remove_pidfd_event 12
1636743144.801 INFO shellous PipeRunner.__aenter__ stepout <PipeRunner 'env|tr'> ex=None
1636743144.802 INFO shellous read_lines stepin
1636743144.803 INFO asyncio <_UnixSubprocessTransport pid=36768 running stdin=<_UnixWritePipeTransport closed fd=13 closed>> exited with return code 0
1636743144.805 INFO shellous Runner._waiter stepout <Runner 'env' pid=36768 exit_code=0> ex=None
1636743144.806 INFO shellous Runner._wait stepout <Runner 'env' pid=36768 exit_code=0> ex=None
1636743144.806 INFO shellous Runner._close stepin <Runner 'env' pid=36768 exit_code=0>
1636743144.809 INFO shellous os.waitpid returned (36769, 0)
1636743144.811 DEBUG shellous _remove_pidfd_event 16
1636743144.812 INFO asyncio <_UnixSubprocessTransport pid=36769 running stdout=<_UnixReadPipeTransport fd=14 polling>> exited with return code 0
1636743144.826 INFO shellous Runner._waiter stepout <Runner 'tr' pid=36769 exit_code=0> ex=None
1636743144.827 INFO asyncio <_UnixReadPipeTransport fd=14 polling> was closed by peer
1636743144.827 INFO shellous Runner._close stepout <Runner 'env' pid=36768 exit_code=0> ex=None
1636743144.833 INFO shellous Runner._finish stepout <Runner 'env' pid=36768 exit_code=0> ex=None
1636743144.834 INFO shellous Runner.__aexit__ stepout <Runner 'env' pid=36768 exit_code=0> ex=None exc_value=None
1636743144.835 INFO shellous PipeRunner.__aexit__ stepin <PipeRunner 'env|tr'> exc_value=None
1636743144.835 INFO shellous PipeRunner._finish stepin <PipeRunner 'env|tr'>
1636743144.836 INFO shellous PipeRunner._wait stepin <PipeRunner 'env|tr'>
1636743144.837 INFO shellous Runner._wait stepout <Runner 'tr' pid=36769 exit_code=0> ex=None
1636743144.838 INFO shellous Runner._close stepin <Runner 'tr' pid=36769 exit_code=0>
1636743144.839 INFO shellous Runner._close stepout <Runner 'tr' pid=36769 exit_code=0> ex=None
1636743144.839 INFO shellous Runner._finish stepout <Runner 'tr' pid=36769 exit_code=0> ex=None
1636743144.840 INFO shellous Runner.__aexit__ stepout <Runner 'tr' pid=36769 exit_code=0> ex=None exc_value=None
1636743144.841 INFO shellous PipeRunner._wait stepout <PipeRunner 'env|tr' results=[Result(output_bytes=b'', exit_code=0, cancelled=False, encoding='utf-8', extra=None), Result(output_bytes=b'', exit_code=0, cancelled=False, encoding='utf-8', extra=None)]> ex=None
1636743144.842 INFO shellous PipeRunner._finish stepout <PipeRunner 'env|tr' results=[Result(output_bytes=b'', exit_code=0, cancelled=False, encoding='utf-8', extra=None), Result(output_bytes=b'', exit_code=0, cancelled=False, encoding='utf-8', extra=None)]> ex=None
1636743144.842 INFO shellous PipeRunner.__aexit__ stepout <PipeRunner 'env|tr' results=[Result(output_bytes=b'', exit_code=0, cancelled=False, encoding='utf-8', extra=None), Result(output_bytes=b'', exit_code=0, cancelled=False, encoding='utf-8', extra=None)]> ex=None exc_value=None
1636743144.843 INFO shellous read_lines stepout ex=GeneratorExit()
PASSED
---------------------------------------------------------------------------------------------------------------------------------- live log teardown ----------------------------------------------------------------------------------------------------------------------------------
1636743144.846 DEBUG shellous watch_pid None
1636743144.847 DEBUG shellous KQWorker stopping <KQueueWorker(Thread-15, started daemon 140513983432448)>
1636743144.849 DEBUG asyncio Close <_UnixSelectorEventLoop running=False closed=False debug=True>

tests/test_shellous.py::test_pipe_breaking_out_of_async_iter ERROR

======================================================================================================================================= ERRORS ========================================================================================================================================
______________________________________________________________________________________________________________ ERROR at teardown of test_pipe_breaking_out_of_async_iter ______________________________________________________________________________________________________________

    def finalizer():
        """Yield again, to finalize."""
    
        async def async_finalizer():
            try:
                await gen_obj.__anext__()
            except StopAsyncIteration:
                pass
            else:
                msg = "Async generator fixture didn't stop."
                msg += "Yield only once."
                raise ValueError(msg)
    
>       loop.run_until_complete(async_finalizer())

.venv/lib/python3.9/site-packages/pytest_asyncio/plugin.py:139: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/usr/lib/python3.9/asyncio/base_events.py:642: in run_until_complete
    return future.result()
.venv/lib/python3.9/site-packages/pytest_asyncio/plugin.py:131: in async_finalizer
    await gen_obj.__anext__()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    @pytest.fixture(autouse=True)
    async def report_orphan_tasks():
        "Make sure that all async tests exit with only a single task running."
    
        # Only run asyncio tests on the main thread. There may be limitations on
        # the childwatcher.
        assert threading.current_thread() is threading.main_thread()
    
        with _check_open_fds():
            yield
            # Close the childwatcher *before* checking for open fd's.
            if sys.platform != "win32":
                cw = asyncio.get_child_watcher()
                if isinstance(cw, DefaultChildWatcher):
                    cw.close()
    
        # Check if any tasks are still running.
        tasks = asyncio.all_tasks()
        if len(tasks) > 1:
            extra_tasks = tasks - {asyncio.current_task()}
>           pytest.fail(f"Orphan tasks still running: {extra_tasks}")
E           Failed: Orphan tasks still running: {<Task pending name='Task-493' coro=<<async_generator_athrow without __name__>()> created at /usr/lib/python3.9/asyncio/events.py:80>}

tests/conftest.py:95: Failed
```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -101,8 +101,8 @@ jobs:
         fi
       env:
         SHELLOUS_CHILDWATCHER_TYPE: "multi"
-    - name: Run Tests with DefaultChildWatcher (MacOS)
-      if: matrix.os == 'macos-11'
+    - name: Run Tests with DefaultChildWatcher (Linux and MacOS)
+      if: matrix.os == 'ubuntu-latest' || matrix.os == 'macos-11'
       run: |
         # Limit to 5 failures for now because GH actions is auto-cancelling...
         $POETRY_RUN python -m pytest -vv -s --log-cli-level=DEBUG --maxfail=5 | ./ci/annotate.awk -v step=DefaultChildWatcher

--- a/shellous/watcher.py
+++ b/shellous/watcher.py
@@ -120,11 +120,8 @@ class KQueueWorker(threading.Thread):
         "Initialize worker variables and start thread."
         super().__init__(daemon=True)
         self._work_queue = queue.Queue()
-        self._client_sock, self._server_sock = self._make_self_pipe()
-        self._server_pids = {}
-        self._fallback_pids = set()
-        self._kqueue = None
-        self._running = True
+        self._client_sock, server_sock = self._make_self_pipe()
+        self._agent = KQueueAgent(self._work_queue, server_sock)
         self.start()
 
     def watch_pid(self, pid, callback, args):
@@ -141,37 +138,62 @@ class KQueueWorker(threading.Thread):
     def run(self):
         "Override Thread.run()."
         LOGGER.debug("KQWorker starting %r", self)
-        self._kqueue = select.kqueue()
         try:
-            self._event_loop()
+            self._agent.event_loop()
         except BaseException as ex:  # pylint: disable=broad-except
             LOGGER.error("KQWorker failed %s ex=%r", self, ex, exc_info=True)
             raise
         finally:
-            self._kqueue.close()
-            self._server_sock.close()
             self._client_sock.close()
             LOGGER.debug("KQWorker stopping %r", self)
 
-    def _event_loop(self):
+    @staticmethod
+    def _make_self_pipe():
+        "Return (client, server) socket pair."
+        client, server = socket.socketpair()
+        client.setblocking(False)
+        server.setblocking(False)
+        return (client, server)
+
+
+class KQueueAgent:
+    "Agent that watches for child exit kqueue event."
+
+    def __init__(self, work_queue, server_sock):
+        "Initialize agent variables."
+        self._work_queue = work_queue
+        self._server_sock = server_sock
+        self._server_pids = {}
+        self._fallback_pids = set()
+        self._kqueue = None
+        self._running = True
+
+    def event_loop(self):
         "Event loop that handles kqueue events."
-        self._add_read_event(self._server_sock.fileno())
-        self._add_signal_event(signal.SIGCHLD)
+        self._kqueue = select.kqueue()
 
-        event_handlers = {
-            select.KQ_FILTER_PROC: self._handle_proc,
-            select.KQ_FILTER_READ: self._handle_read,
-            select.KQ_FILTER_SIGNAL: self._handle_signal,
-        }
+        try:
+            self._add_read_event(self._server_sock.fileno())
+            self._add_signal_event(signal.SIGCHLD)
 
-        while self._running:
-            pending = self._kqueue.control(None, 10, 10)
-            for event in pending:
-                handler = event_handlers.get(event.filter)
-                if handler:
-                    handler(event)
-                else:
-                    LOGGER.debug("_event_loop: unknown kevent: %r", event)
+            event_handlers = {
+                select.KQ_FILTER_PROC: self._handle_proc,
+                select.KQ_FILTER_READ: self._handle_read,
+                select.KQ_FILTER_SIGNAL: self._handle_signal,
+            }
+
+            while self._running:
+                pending = self._kqueue.control(None, 10, 10)
+                for event in pending:
+                    handler = event_handlers.get(event.filter)
+                    if handler:
+                        handler(event)
+                    else:
+                        LOGGER.debug("_event_loop: unknown kevent: %r", event)
+
+        finally:
+            self._kqueue.close()
+            self._server_sock.close()
 
     def _handle_proc(self, event):
         "Handle KQ_FILTER_PROC event."
@@ -179,7 +201,7 @@ class KQueueWorker(threading.Thread):
 
     def _handle_read(self, _event):
         "Handle KQ_FILTER_READ event."
-        self._drain_server_sock()
+        _drain(self._server_sock)
         self._check_queue()
 
     def _handle_signal(self, event):
@@ -206,17 +228,6 @@ class KQueueWorker(threading.Thread):
             LOGGER.error("_check_pid: unregistered pid: %r", pid)
 
         return False
-
-    def _drain_server_sock(self):
-        "Ref: asyncio/selector_events.py#L117."
-        while True:
-            try:
-                if not self._server_sock.recv(4096):
-                    break
-            except InterruptedError:
-                continue
-            except BlockingIOError:
-                break
 
     def _check_queue(self):
         "Check the work queue for new work requests."
@@ -313,10 +324,16 @@ class KQueueWorker(threading.Thread):
         except Exception as ex:  # pylint: disable=broad-except
             LOGGER.error("_add_signal_event: ex=%r", ex)
 
-    @staticmethod
-    def _make_self_pipe():
-        "Return (client, server) socket pair."
-        client, server = socket.socketpair()
-        client.setblocking(False)
-        server.setblocking(False)
-        return (client, server)
+
+def _drain(sock):
+    """Utility function to drain a socket.
+
+    Ref: asyncio/selector_events.py#L117."""
+    while True:
+        try:
+            if not sock.recv(4096):
+                break
+        except InterruptedError:
+            continue
+        except BlockingIOError:
+            break

--- a/shellous/watcher.py
+++ b/shellous/watcher.py
@@ -114,8 +114,6 @@ class KQueueWorker(threading.Thread):
 
     """
 
-    # pylint: disable=no-member
-
     def __init__(self):
         "Initialize worker variables and start thread."
         super().__init__(daemon=True)
@@ -158,6 +156,8 @@ class KQueueWorker(threading.Thread):
 
 class KQueueAgent:
     "Agent that watches for child exit kqueue event."
+
+    # pylint: disable=no-member
 
     def __init__(self, work_queue, server_sock):
         "Initialize agent variables."

--- a/shellous/watcher.py
+++ b/shellous/watcher.py
@@ -159,11 +159,11 @@ class KQueueAgent:
 
     # pylint: disable=no-member
 
-    def __init__(self, work_queue, server_sock):
+    def __init__(self, work_queue, wakeup_sock):
         "Initialize agent variables."
         self._work_queue = work_queue
-        self._server_sock = server_sock
-        self._server_pids = {}
+        self._wakeup_sock = wakeup_sock
+        self._active_pids = {}
         self._fallback_pids = set()
         self._kqueue = None
         self._running = True
@@ -173,7 +173,7 @@ class KQueueAgent:
         self._kqueue = select.kqueue()
 
         try:
-            self._add_read_event(self._server_sock.fileno())
+            self._add_read_event(self._wakeup_sock.fileno())
             self._add_signal_event(signal.SIGCHLD)
 
             event_handlers = {
@@ -193,7 +193,7 @@ class KQueueAgent:
 
         finally:
             self._kqueue.close()
-            self._server_sock.close()
+            self._wakeup_sock.close()
 
     def _handle_proc(self, event):
         "Handle KQ_FILTER_PROC event."
@@ -201,7 +201,7 @@ class KQueueAgent:
 
     def _handle_read(self, _event):
         "Handle KQ_FILTER_READ event."
-        _drain(self._server_sock)
+        _drain(self._wakeup_sock)
         self._check_queue()
 
     def _handle_signal(self, event):
@@ -211,8 +211,8 @@ class KQueueAgent:
                 self._fallback_pids.remove(pid)
 
     def _check_pid(self, pid):
-        """Called when a process exits."""
-        info = self._server_pids.get(pid, None)
+        """Called when a process exits. Returns true if process exited."""
+        info = self._active_pids.get(pid, None)
         if info:
             callback, args = info
             status = wait_pid(pid)
@@ -222,7 +222,7 @@ class KQueueAgent:
             else:
                 # Invoke callback function here.
                 callback(pid, status, *args)
-                self._server_pids.pop(pid)
+                self._active_pids.pop(pid)
                 return True
         else:
             LOGGER.error("_check_pid: unregistered pid: %r", pid)
@@ -244,27 +244,19 @@ class KQueueAgent:
             pass
 
     def _monitor_pid(self, pid, callback, args):
-        """Add pid and info to our list of pid's to be monitored.
-
-        Return true if we successfully added the pid to kqueue.
-        """
-        if pid in self._server_pids:
+        """Add pid and info to our list of pid's to be monitored."""
+        if pid in self._active_pids:
             LOGGER.warning("_monitor_pid: already monitored? pid=%r", pid)
 
-        self._server_pids[pid] = (callback, args)
-        if self._add_proc_event(pid):
-            return True
-
-        return self._monitor_pid_fallback(pid)
+        self._active_pids[pid] = (callback, args)
+        if not self._add_proc_event(pid):
+            self._monitor_pid_fallback(pid)
 
     def _monitor_pid_fallback(self, pid):
         "Called when _monitor_pid fails to add a kevent."
-        if self._check_pid(pid):
-            return True
-
-        LOGGER.warning("_monitor_pid_fallback: adding fallback pid=%r", pid)
-        self._fallback_pids.add(pid)
-        return False
+        if not self._check_pid(pid):
+            LOGGER.warning("_monitor_pid_fallback: adding fallback pid=%r", pid)
+            self._fallback_pids.add(pid)
 
     def _add_proc_event(self, pid):
         """Add kevent that monitors for process exit.

--- a/shellous/watcher.py
+++ b/shellous/watcher.py
@@ -37,7 +37,7 @@ class DefaultChildWatcher(asyncio.AbstractChildWatcher):
         self._worker = None
 
         # Install dummy SIGCHLD signal handler to make EV_FILTER_SIGNAL work.
-        signal.signal(signal.SIGCHLD, lambda *args: None)
+        signal.signal(signal.SIGCHLD, signal.SIG_DFL)
 
     def add_child_handler(self, pid, callback, *args):
         """Register a new child handler.
@@ -96,7 +96,7 @@ class DefaultChildWatcher(asyncio.AbstractChildWatcher):
 
         return self
 
-    def __exit__(self, a, b, c):
+    def __exit__(self, *_args):
         """Exit the watcher's context"""
 
 
@@ -334,6 +334,10 @@ class EPollAgent:
         self._pidfds = {}  # pidfd -> pid
         self._epoll = None
         self._running = True
+
+        assert (
+            signal.getsignal(signal.SIGCHLD) != signal.SIG_IGN
+        ), "pidfd does not work when SIGCHLD set to SIG_IGN"
 
     def event_loop(self):
         "Event loop that handles kqueue events."

--- a/shellous/watcher.py
+++ b/shellous/watcher.py
@@ -414,14 +414,21 @@ class EPollAgent:
 
         self._active_pids[pid] = (callback, args)
         if not exists:
-            self._add_pid_event(pid)
+            if not self._add_pid_event(pid):
+                # If pidfd_open failed, check if the process already exited.
+                self._check_pid(pid)
 
     def _add_pid_event(self, pid):
-        "Add epoll that monitors for process exit."
-        pidfd = os.pidfd_open(pid, 0)
+        "Add epoll that monitors for process exit. Return True if successful."
+        try:
+            pidfd = os.pidfd_open(pid, 0)
+        except ProcessLookupError:
+            return False
+
         self._add_read_event(pidfd)
         assert pidfd not in self._pidfds
         self._pidfds[pidfd] = pid
+        return True
 
     def _add_read_event(self, fdesc):
         "Add epoll that monitors for file descriptor wakeup."

--- a/shellous/watcher.py
+++ b/shellous/watcher.py
@@ -422,7 +422,8 @@ class EPollAgent:
         "Add epoll that monitors for process exit. Return True if successful."
         try:
             pidfd = os.pidfd_open(pid, 0)
-        except ProcessLookupError:
+        except ProcessLookupError as ex:
+            LOGGER.error("pidfd_open(%r) failed: ex=%r", pid, ex)
             return False
 
         self._add_read_event(pidfd)

--- a/shellous/watcher.py
+++ b/shellous/watcher.py
@@ -36,8 +36,11 @@ class DefaultChildWatcher(asyncio.AbstractChildWatcher):
         self._lock = threading.Lock()  # guard `self`
         self._worker = None
 
-        # Install dummy SIGCHLD signal handler to make EV_FILTER_SIGNAL work.
-        signal.signal(signal.SIGCHLD, signal.SIG_DFL)
+        # kqueue/pidfd do not work if SIGCHLD is set to SIG_IGN in the
+        # signal table. On unix systems, the default action for SIGCHLD is to
+        # discard the signal; this is compatible and does what we want.
+        if signal.getsignal(signal.SIGCHLD) == signal.SIG_IGN:
+            signal.signal(signal.SIGCHLD, signal.SIG_DFL)
 
     def add_child_handler(self, pid, callback, *args):
         """Register a new child handler.
@@ -171,6 +174,10 @@ class KQueueAgent:
         self._fallback_pids = set()
         self._kqueue = None
         self._running = True
+
+        assert (
+            signal.getsignal(signal.SIGCHLD) != signal.SIG_IGN
+        ), "kqueue does not work when SIGCHLD set to SIG_IGN"
 
     def event_loop(self):
         "Event loop that handles kqueue events."

--- a/shellous/watcher.py
+++ b/shellous/watcher.py
@@ -21,15 +21,13 @@ else:
 
 
 class DefaultChildWatcher(asyncio.AbstractChildWatcher):
-    """Uses kqueue to monitor for exiting child processes.
+    """Uses kqueue/pidfd to monitor for exiting child processes.
 
     Design Goals:
       1. Independent of any running event loop.
       2. Zero-cost until used.
 
     Cost: 3 file descriptors and 1 thread.
-
-    A DefaultChildWatcher must be created on the main thread.
     """
 
     def __init__(self):
@@ -38,9 +36,9 @@ class DefaultChildWatcher(asyncio.AbstractChildWatcher):
 
         # kqueue/pidfd do not work if SIGCHLD is set to SIG_IGN in the
         # signal table. On unix systems, the default action for SIGCHLD is to
-        # discard the signal; this is compatible and does what we want.
+        # discard the signal; SIG_DFL is compatible and does what we want.
         if signal.getsignal(signal.SIGCHLD) == signal.SIG_IGN:
-            signal.signal(signal.SIGCHLD, signal.SIG_DFL)
+            raise RuntimeError("SIGCHLD cannot be set to SIG_IGN")
 
     def add_child_handler(self, pid, callback, *args):
         """Register a new child handler.
@@ -395,6 +393,7 @@ class EPollAgent:
                 LOGGER.debug("_check_pid: process still running pid=%r", pid)
             else:
                 # Invoke callback function here.
+                LOGGER.debug("_check_pid: Callback for pid %r, status %r", pid, status)
                 callback(pid, status, *args)
                 self._active_pids.pop(pid)
                 return True
@@ -432,6 +431,11 @@ class EPollAgent:
     def _add_pid_event(self, pid):
         "Add epoll that monitors for process exit. Return True if successful."
         try:
+            # FIXME: At the moment, this design has the pidfd_open done in
+            # another thread, after a possible delay. The call should be done
+            # ASAP after launching the process. If we send a SIGTERM to the
+            # process before calling pidfd_open, I have seen this pidfd_open
+            # fail with ProcessLookupError.
             pidfd = os.pidfd_open(pid, 0)
         except ProcessLookupError as ex:
             LOGGER.error("pidfd_open(%r) failed: ex=%r", pid, ex)

--- a/shellous/watcher.py
+++ b/shellous/watcher.py
@@ -1,6 +1,7 @@
 "Implements DefaultChildWatcher."
 
 import asyncio
+import os
 import queue
 import select
 import signal
@@ -119,7 +120,10 @@ class KQueueWorker(threading.Thread):
         super().__init__(daemon=True)
         self._work_queue = queue.Queue()
         self._client_sock, server_sock = self._make_self_pipe()
-        self._agent = KQueueAgent(self._work_queue, server_sock)
+        if sys.platform == "linux":
+            self._agent = EPollAgent(self._work_queue, server_sock)
+        else:
+            self._agent = KQueueAgent(self._work_queue, server_sock)
         self.start()
 
     def watch_pid(self, pid, callback, args):
@@ -315,6 +319,121 @@ class KQueueAgent:
 
         except Exception as ex:  # pylint: disable=broad-except
             LOGGER.error("_add_signal_event: ex=%r", ex)
+
+
+class EPollAgent:
+    "Agent that watches for child exit epoll event."
+
+    # pylint: disable=no-member
+
+    def __init__(self, work_queue, wakeup_sock):
+        "Initialize agent variables."
+        self._work_queue = work_queue
+        self._wakeup_sock = wakeup_sock
+        self._active_pids = {}  # pid -> (callback, args)
+        self._pidfds = {}  # pidfd -> pid
+        self._epoll = None
+        self._running = True
+
+    def event_loop(self):
+        "Event loop that handles kqueue events."
+        self._epoll = select.epoll()
+
+        try:
+            self._add_read_event(self._wakeup_sock.fileno())
+
+            while self._running:
+                pending = self._epoll.poll()
+                # Each event is a 2-tuple (fd, events):
+                for fd, events in pending:
+                    assert events == select.EPOLLIN
+                    self._handle_read(fd)
+
+        finally:
+            # FIXME need to close all pidfds...
+            self._epoll.close()
+            self._wakeup_sock.close()
+
+    def _handle_read(self, fdesc):
+        "Handle epoll read event."
+
+        if fdesc == self._wakeup_sock.fileno():
+            _drain(self._wakeup_sock)
+            self._check_queue()
+        else:
+            self._handle_pidfd(fdesc)
+
+    def _handle_pidfd(self, pidfd):
+        "Handle epoll pidfd event."
+        pid = self._pidfds.get(pidfd)
+        if pid is None:
+            LOGGER.warning("Unknown pidfd %r", pidfd)
+            return
+
+        if self._check_pid(pid):
+            self._remove_pidfd_event(pidfd)
+
+    def _check_pid(self, pid):
+        """Called when a process exits. Returns true if process exited."""
+        info = self._active_pids.get(pid, None)
+        if info:
+            callback, args = info
+            status = wait_pid(pid)
+            if status is None:
+                # Process is still running.
+                LOGGER.debug("_check_pid: process still running pid=%r", pid)
+            else:
+                # Invoke callback function here.
+                callback(pid, status, *args)
+                self._active_pids.pop(pid)
+                return True
+        else:
+            LOGGER.error("_check_pid: unregistered pid: %r", pid)
+
+        return False
+
+    def _check_queue(self):
+        "Check the work queue for new work requests."
+        try:
+            while True:
+                pid, callback, args = self._work_queue.get_nowait()
+                if pid is None:
+                    self._running = False
+                    break
+
+                self._monitor_pid(pid, callback, args)
+
+        except queue.Empty:
+            pass
+
+    def _monitor_pid(self, pid, callback, args):
+        """Add pid and info to our list of pid's to be monitored."""
+        exists = pid in self._active_pids
+        if exists:
+            LOGGER.warning("_monitor_pid: already monitored? pid=%r", pid)
+
+        self._active_pids[pid] = (callback, args)
+        if not exists:
+            self._add_pid_event(pid)
+
+    def _add_pid_event(self, pid):
+        "Add epoll that monitors for process exit."
+        pidfd = os.pidfd_open(pid, 0)
+        self._add_read_event(pidfd)
+        assert pidfd not in self._pidfds
+        self._pidfds[pidfd] = pid
+
+    def _add_read_event(self, fdesc):
+        "Add epoll that monitors for file descriptor wakeup."
+        LOGGER.debug("_add_read_event %r", fdesc)
+        self._epoll.register(fdesc, select.EPOLLIN)
+
+    def _remove_pidfd_event(self, fdesc):
+        "Remove epoll that monitors for process exit."
+        LOGGER.debug("_remove_pidfd_event %r", fdesc)
+        del self._pidfds[fdesc]
+        self._epoll.unregister(fdesc)
+        os.close(fdesc)
 
 
 def _drain(sock):

--- a/shellous/watcher.py
+++ b/shellous/watcher.py
@@ -10,7 +10,7 @@ import sys
 import threading
 
 from shellous.log import LOGGER
-from shellous.util import wait_pid
+from shellous.util import close_fds, wait_pid
 
 assert sys.platform != "win32"
 
@@ -359,9 +359,9 @@ class EPollAgent:
                     self._handle_read(fd)
 
         finally:
-            # FIXME need to close all pidfds...
             self._epoll.close()
             self._wakeup_sock.close()
+            close_fds(list(self._pidfds.keys()))
 
     def _handle_read(self, fdesc):
         "Handle epoll read event."

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -4,6 +4,7 @@ import asyncio
 import contextlib
 import functools
 import gc
+import logging
 import os
 import re
 import signal
@@ -88,14 +89,22 @@ async def report_orphan_tasks():
             if isinstance(cw, DefaultChildWatcher):
                 cw.close()
 
-    # Check if any tasks are still running.
-    tasks = asyncio.all_tasks()
-    if len(tasks) > 1:
-        extra_tasks = tasks - {asyncio.current_task()}
-        pytest.fail(f"Orphan tasks still running: {extra_tasks}")
+    # Check if any other tasks are still running. Ignore the current task.
+    extra_tasks = asyncio.all_tasks() - {asyncio.current_task()}
 
-    # We expect the only task to be the current task.
-    assert tasks.pop() is asyncio.current_task()
+    # Check if any running tasks are related to async generators. If so, yield
+    # time to get them to exit and update `extra_tasks`.
+    agen_tasks = {
+        task
+        for task in extra_tasks
+        if "<async_generator_athrow without __name__>" in repr(task)
+    }
+    if agen_tasks:
+        await asyncio.sleep(0)
+        extra_tasks = asyncio.all_tasks() - {asyncio.current_task()}
+
+    if extra_tasks:
+        pytest.fail(f"Orphan tasks still running: {extra_tasks}")
 
     # Garbage collect here to flush out warnings from __del__ methods
     # while loop is still running.

--- a/tests/unix/test_threads.py
+++ b/tests/unix/test_threads.py
@@ -12,6 +12,7 @@ import shellous
 from shellous.log import log_method
 
 if sys.platform != "win32":
+    from shellous.watcher import DefaultChildWatcher
     from tests.conftest import PatchedMultiLoopChildWatcher
 
 unix_only = pytest.mark.skipif(sys.platform == "win32", reason="Unix")
@@ -91,6 +92,8 @@ def run_in_thread(child_watcher_name="ThreadedChildWatcher"):
         def _wrap(*args, **kwargs):
             if child_watcher_name == "MultiLoopChildWatcher":
                 child_watcher = PatchedMultiLoopChildWatcher()
+            elif child_watcher_name == "DefaultChildWatcher":
+                child_watcher = DefaultChildWatcher()
             else:
                 child_watcher = getattr(asyncio, child_watcher_name)()
             asyncio.set_child_watcher(child_watcher)
@@ -133,6 +136,7 @@ _CHILD_WATCHER_MAP = {
     "safe": "SafeChildWatcher",
     "pidfd": "PidfdChildWatcher",
     "multi": "MultiLoopChildWatcher",
+    "default": "DefaultChildWatcher",
 }
 
 _CW_TYPE = os.environ.get("SHELLOUS_CHILDWATCHER_TYPE", "<unset>")

